### PR TITLE
Create build source citation function

### PR DESCRIPTION
**Pull request by LienReyserhove**
In this PR, I would like to receive some feedback on the function (my first!) I wrote.

### CODE REVIEW

**Comment by LienReyserhove** [comment]
I see now that all the commits for the DwC mapping are in this PR, :disappointed:

File: src/dwc_mapping.Rmd [hunk 1]
```diff
@@ -348,6 +348,55 @@ write_csv(taxon, here("data", "processed", "taxon.csv"), na = "")
...
+  checklist_citation <- 
+    checklists %>% 
+    select(datasetKey, citation)
+  
+  # Join the extension and source_citation by their datasetKey:
+  
+  extension <-
+    extension %>% 
+    left_join(
+      checklist_citation,
```
Comment (peterdesmet) [inline src/dwc_mapping.Rmd]: I would use `select(checklists, datasetKey, citation)` here and remove the step before (lines 363 to 367).

File: src/dwc_mapping.Rmd [hunk 2]
```diff
@@ -348,6 +348,55 @@ write_csv(taxon, here("data", "processed", "taxon.csv"), na = "")
...
+
+All records in the extensions will be documented with a source, i.e. the link to the specific checklist taxon that was used for the specific piece of information. We build this function here first.
+
+```{r}
+build_source_citation <- function(extension, datasetKey_extension){
+  
+  # Select the citation of the source checlist: 
+  
+  checklist_citation <- 
+    checklists %>%
```
Comment (peterdesmet) [inline src/dwc_mapping.Rmd]: This variable `checklists` should be passed  to the function

File: src/dwc_mapping.Rmd [hunk 3]
```diff
@@ -348,6 +348,55 @@ write_csv(taxon, here("data", "processed", "taxon.csv"), na = "")
...
+
+```
+
+
+# Create source citation function
+
+All records in the extensions will be documented with a source, i.e. the link to the specific checklist taxon that was used for the specific piece of information. We build this function here first.
+
+```{r}
+build_source_citation <- function(extension, datasetKey_extension){
```
Comment (peterdesmet) [inline src/dwc_mapping.Rmd]: I would expect the function to be `function(extension, checklists)`

File: src/dwc_mapping.Rmd [hunk 4]
```diff
@@ -348,6 +348,55 @@ write_csv(taxon, here("data", "processed", "taxon.csv"), na = "")
...
+    checklists %>% 
+    select(datasetKey, citation)
+  
+  # Join the extension and source_citation by their datasetKey:
+  
+  extension <-
+    extension %>% 
+    left_join(
+      checklist_citation,
+      by = setNames("datasetKey", datasetKey_extension)
```
Comment (peterdesmet) [inline src/dwc_mapping.Rmd]: I would expect `on = datasetKey`

File: src/dwc_mapping.Rmd (hunk 3 above)
Comment (peterdesmet) [reply src/dwc_mapping.Rmd]: Ah right, you also want to provide the _name_ of the `datasetKey` column in the extension. On the other hand, you already assume that the other names (e.g. `input_scientificName`) are written as such, so I would just assume it here as well. For a proper function (expected to be used by others), you have to offer more freedom (let the user provide their column names), but for a quick helper function like here, I would just make a couple of assumptions (e.g. that the column is named `input_datasetKey`). So it should remain: `function(extension, checklists)`.

File: src/dwc_mapping.Rmd (hunk 4 above)
Comment (peterdesmet) [reply src/dwc_mapping.Rmd]: Disregard, should be `setNames("input_datasetKey", "datasetKey")`.

File: src/dwc_mapping.Rmd [hunk 5]
```diff
@@ -348,6 +348,55 @@ write_csv(taxon, here("data", "processed", "taxon.csv"), na = "")
...
+  extension <-
+    extension %>% 
+    mutate(source = paste(
+      "https://www.gbif.org/species/",
+      input_taxonKey,
+      ":",
+      input_scientificName,
+      "in",
+      citation,
+      sep = " "
```
Comment (peterdesmet) [inline src/dwc_mapping.Rmd]: This is the default for paste, so no need to provide. But you actually want no spaces, so would use `paste0()`

File: src/dwc_mapping.Rmd [hunk 6]
```diff
@@ -348,6 +348,55 @@ write_csv(taxon, here("data", "processed", "taxon.csv"), na = "")
...
+  
+  # Generate the field `source` by combining `taxonKey`, `scientificName` and `citation`. This is an example of how the final source should look like: 
+  # https://www.gbif.org/species/141264581: Nymphaea marliacea Marliac in Verloove F, Groom Q, Brosens D, Desmet P, Reyserhove L (2018). Manual of the Alien Plants of Belgium. Version 1.7. Botanic Garden Meise. Checklist dataset https://doi.org/10.15468/wtda1m
+  
+  extension <-
+    extension %>% 
+    mutate(source = paste(
+      "https://www.gbif.org/species/",
+      input_taxonKey,
+      ":",
```
Comment (peterdesmet) [inline src/dwc_mapping.Rmd]: `: ` (including space)

File: src/dwc_mapping.Rmd [hunk 7]
```diff
@@ -348,6 +348,55 @@ write_csv(taxon, here("data", "processed", "taxon.csv"), na = "")
...
+      input_scientificName,
+      "in",
```
Comment (peterdesmet) [inline src/dwc_mapping.Rmd]: ` in ` (including two spaces)

File: src/dwc_mapping.Rmd [hunk 8]
```diff
@@ -348,6 +348,55 @@ write_csv(taxon, here("data", "processed", "taxon.csv"), na = "")
...
+    extension %>% 
+    left_join(
+      checklist_citation,
+      by = setNames("datasetKey", datasetKey_extension)
+    ) 
+  
+  # Generate the field `source` by combining `taxonKey`, `scientificName` and `citation`. This is an example of how the final source should look like: 
+  # https://www.gbif.org/species/141264581: Nymphaea marliacea Marliac in Verloove F, Groom Q, Brosens D, Desmet P, Reyserhove L (2018). Manual of the Alien Plants of Belgium. Version 1.7. Botanic Garden Meise. Checklist dataset https://doi.org/10.15468/wtda1m
+  
+  extension <-
```
Comment (peterdesmet) [inline src/dwc_mapping.Rmd]: You don't really have to assign this back to `extension`. The result of your code will be returned by the function.

File: src/dwc_mapping.Rmd (hunk 3 above)
Comment (LienReyserhove) [reply src/dwc_mapping.Rmd]: Ok, I agree. In that case, I would also eliminate `checklists` from the function, as this dataframe is reused each time and is fixed. So the only argument in the function would be `extension`

Comment (peterdesmet) [reply src/dwc_mapping.Rmd]: No, you should assume a function works in its own space: you have to pass all the information it needs.

File: src/dwc_mapping.Rmd [hunk 9]
```diff
@@ -263,84 +347,125 @@ Save to CSV:
...
+  # Generate the field `source` by combining `taxonKey`, `scientificName` and `citation`. This is an example of how the final source should look like: 
+  # https://www.gbif.org/species/141264581: Nymphaea marliacea Marliac in Verloove F, Groom Q, Brosens D, Desmet P, Reyserhove L (2018). Manual of the Alien Plants of Belgium. Version 1.7. Botanic Garden Meise. Checklist dataset https://doi.org/10.15468/wtda1m
+  
+    extension %>% 
+    mutate(source = paste0(
+      "https://www.gbif.org/species/",
+      input_taxonKey,
+      ": ",
+      input_scientificName,
+      "in  ",
```
Comment (peterdesmet) [inline src/dwc_mapping.Rmd]: I meant a space before and after 😄

Comment (LienReyserhove) [reply src/dwc_mapping.Rmd]: :smile: I thought it was a bit odd but in these things, I just trust your opinion ;-)

**Comment by LienReyserhove** [comment]
I integrated the changes, thx!

**Pull request merged**